jerry-util/arg-validator/arg-validator.h

```diff
@@ -0,0 +1,94 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRY_ARG_VALIDATOR_H
+#define JERRY_ARG_VALIDATOR_H
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "jerryscript.h"
+
+typedef struct jerry_arg_t jerry_arg_t;
+typedef struct jerry_arg_iterator_t jerry_arg_iterator_t;
+
+/**
+ * Signature of the validator funtion.
+ */
+typedef jerry_value_t (*jerry_arg_func_t) (jerry_arg_iterator_t *iterator_p);
+
+/**
+ * The structure of the jerry argument in validator.
+ * Each native argument should have one corresponding jerry_arg_t item.
+ */
+struct jerry_arg_t
+{
+    jerry_arg_func_t func; /**< the validator function for the native argument */
+    void *dest; /**< point to the native argument */
+    /** The extra infomation for validator */
+    union
+    {
+        void *obj_p; /**< the pointer to native handle's expected type info */
+        jerry_length_t length; /**< the length of the string */
+    } info;
+};
+
+/**
+ * The argument iteration used for validator funtions.
+ */
+struct jerry_arg_iterator_t
+{
+    const jerry_arg_t *c_arg_p; /**< points to the current jerry_arg_t */
+    const jerry_value_t *js_arg_p; /**< points to the current js argument */
+    const jerry_length_t  js_arg_cnt; /**< the total count of the js arguments */
+    jerry_length_t  js_arg_index; /**< the current index of the js argument */
+};
```

Comment:
That is because `jerry_arg_t` is used in the definition of `jerry_iteratior_t`,  `jerry_arg_func_t` is used in the definition of `jerry_arg_t`, and  `jerry_iteratior_t` is used in the definition of `jerry_arg_func_t`.
I had to declare those struct first.